src/v/coproc/script_context_router.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/partition.h"
+#include "model/fundamental.h"
+
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/flat_hash_map.h>
+
+namespace coproc {
+
+/// Read state for a single input topic, unique and unshared
+struct read_context {
+    /// Where to begin reading from, determined by user provided option \ref
+    /// topic_ingestion_policy, earliest, latest, or stored
+    model::offset absolute_start{};
+    /// Highest offset of current in-progress read
+    model::offset last_read{};
+    /// Highest offset that all materialized topics have observed
+    model::offset last_acked{};
+    /// Pointer to an input topic
+    ss::lw_shared_ptr<cluster::partition> input;
+};
+
+/// Write state for topics that have been created stemming from an associated
+/// source, the input begint the pointer in \ref read_context
+struct write_context {
+    /// Where 'key' is the materialized_ntp and offset represets an ack of the
+    /// read off of the respective input topic
+    using offsets_t = absl::flat_hash_map<model::ntp, model::offset>;
+
+    offsets_t offsets;
+
+    /// Returns the min offset in the collection, O(n) runtime
+    model::offset min_offset() const {
```

Comment:
`std::min_element`?

Reply:
Not sure if that will work since *itr in a map is `Map::value_type` or `std::pair<key, value>`. When I want to compare values themselves not the pair

Reply:
It's actually a `std::pair<const key, value>`

There's a version of `min_element that takes a predicate, or you could create a `transform_iterator` that returns `.second`.